### ISSUE TITLE
perlperf.pod wrong example

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#10346]
Migrated from [rt.perl.org#74650](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=74650) (status was 'stalled')

Searchable as RT74650$

**Comment by p5pRT** [comment]
### From occitan@esperanto.org

In the dereference benchmark you make exactly the mistake you warn of 
before&#8203;: you change several factors at once and show an overdrawn 
conclusion\.  While refactoring out the common $ref\->\{ref\} to access it 
only once\, you unnecessarily also copy both strings to a local 
variable\.  It is this second action that drives the time up\, if you 
leave that out and run timethese several times your findings are still 
right\, but far less\.  If however you need a third or more values \(which 
I added of the same length\)\, dereferencing the outer hash only once 
becomes a clear winner&#8203;:

&nbsp;    timethese\(1000000\, \{
&nbsp;        '2direct'      => sub \{
&nbsp;            my $x = $ref\->\{ref\}\{\_myscore\} \. $ref\->\{ref\}\{\_yourscore\} ;
&nbsp;        \}\,
&nbsp;        '2wrong'       => sub \{
&nbsp;            my $ref = $ref\->\{ref\};
&nbsp;            my $myscore = $ref\->\{\_myscore\};        \# why?
&nbsp;            my $yourscore = $ref\->\{\_yourscore\};    \# why?
&nbsp;            my $x = $myscore \. $yourscore;
&nbsp;        \}\,
&nbsp;        '2dereference' => sub \{
&nbsp;            my $ref = $ref\->\{ref\};
&nbsp;            my $x = $ref\->\{\_myscore\} \. $ref\->\{\_yourscore\};
&nbsp;        \}\,
&nbsp;        '3direct'      => sub \{
&nbsp;            my $x = $ref\->\{ref\}\{\_myscore\} \. $ref\->\{ref\}\{\_yourscore\} \.
&nbsp;                $ref\->\{ref\}\{\_ourscore\};
&nbsp;        \}\,
&nbsp;        '3dereference' => sub \{
&nbsp;            my $ref = $ref\->\{ref\};
&nbsp;            my $x = $ref\->\{\_myscore\} \. $ref\->\{\_yourscore\} \.
&nbsp;                $ref\->\{\_ourscore\};
&nbsp;        \}\,
&nbsp;    \}\);

Btw\. I have left out the intermediate arrow\, making it less of an 
eyesore and easier to understand as a two dimensional hash\.

coralament / best Grötens / liebe Grüße / best regards / elkorajn salutojn
Daniel Pfeiffer

\-\- 
lerne / learn / apprends / lär dig / ucz się    Esperanto&#8203;:
&nbsp;                    http&#8203;://lernu\.net  /  http&#8203;://ikurso\.net

**Comment by p5pRT** [comment]
### From @richardfoley

Hi Daniel\,

Thanks very much for your feedback \- anticipating your patch to fix ?\-\)

\-\-
Richard Foley
Ciao \- shorter than aufwiedersehen

http&#8203;://www\.rfi\.net/

On Sunday 25 April 2010 15&#8203;:20&#8203;:22 Daniel Pfeiffer wrote&#8203;:
> In the dereference benchmark you make exactly the mistake you warn of
> before&#8203;: you change several factors at once and show an overdrawn
> conclusion\.  While refactoring out the common $ref\->\{ref\} to access it
> only once\, you unnecessarily also copy both strings to a local
> variable\.  It is this second action that drives the time up\, if you
> leave that out and run timethese several times your findings are still
> right\, but far less\.  If however you need a third or more values \(which
> I added of the same length\)\, dereferencing the outer hash only once
> becomes a clear winner&#8203;:
>
>     timethese\(1000000\, \{
>         '2direct'      => sub \{
>             my $x = $ref\->\{ref\}\{\_myscore\} \. $ref\->\{ref\}\{\_yourscore\} ;
>         \}\,
>         '2wrong'       => sub \{
>             my $ref = $ref\->\{ref\};
>             my $myscore = $ref\->\{\_myscore\};        \# why?
>             my $yourscore = $ref\->\{\_yourscore\};    \# why?
>             my $x = $myscore \. $yourscore;
>         \}\,
>         '2dereference' => sub \{
>             my $ref = $ref\->\{ref\};
>             my $x = $ref\->\{\_myscore\} \. $ref\->\{\_yourscore\};
>         \}\,
>         '3direct'      => sub \{
>             my $x = $ref\->\{ref\}\{\_myscore\} \. $ref\->\{ref\}\{\_yourscore\} \.
>                 $ref\->\{ref\}\{\_ourscore\};
>         \}\,
>         '3dereference' => sub \{
>             my $ref = $ref\->\{ref\};
>             my $x = $ref\->\{\_myscore\} \. $ref\->\{\_yourscore\} \.
>                 $ref\->\{\_ourscore\};
>         \}\,
>     \}\);
>
> Btw\. I have left out the intermediate arrow\, making it less of an
> eyesore and easier to understand as a two dimensional hash\.
>
> coralament / best Grötens / liebe Grüße / best regards / elkorajn salutojn
> Daniel Pfeiffer

**Comment by p5pRT** [comment]
### From @richardfoley

Hi Daniel\,

Thanks very much for your feedback \- anticipating your patch to fix ?\-\)

\-\-
Richard Foley
Ciao \- shorter than aufwiedersehen

http&#8203;://www\.rfi\.net/

On Sunday 25 April 2010 15&#8203;:20&#8203;:22 Daniel Pfeiffer wrote&#8203;:
> In the dereference benchmark you make exactly the mistake you warn of
> before&#8203;: you change several factors at once and show an overdrawn
> conclusion\.  While refactoring out the common $ref\->\{ref\} to access it
> only once\, you unnecessarily also copy both strings to a local
> variable\.  It is this second action that drives the time up\, if you
> leave that out and run timethese several times your findings are still
> right\, but far less\.  If however you need a third or more values \(which
> I added of the same length\)\, dereferencing the outer hash only once
> becomes a clear winner&#8203;:
>
>     timethese\(1000000\, \{
>         '2direct'      => sub \{
>             my $x = $ref\->\{ref\}\{\_myscore\} \. $ref\->\{ref\}\{\_yourscore\} ;
>         \}\,
>         '2wrong'       => sub \{
>             my $ref = $ref\->\{ref\};
>             my $myscore = $ref\->\{\_myscore\};        \# why?
>             my $yourscore = $ref\->\{\_yourscore\};    \# why?
>             my $x = $myscore \. $yourscore;
>         \}\,
>         '2dereference' => sub \{
>             my $ref = $ref\->\{ref\};
>             my $x = $ref\->\{\_myscore\} \. $ref\->\{\_yourscore\};
>         \}\,
>         '3direct'      => sub \{
>             my $x = $ref\->\{ref\}\{\_myscore\} \. $ref\->\{ref\}\{\_yourscore\} \.
>                 $ref\->\{ref\}\{\_ourscore\};
>         \}\,
>         '3dereference' => sub \{
>             my $ref = $ref\->\{ref\};
>             my $x = $ref\->\{\_myscore\} \. $ref\->\{\_yourscore\} \.
>                 $ref\->\{\_ourscore\};
>         \}\,
>     \}\);
>
> Btw\. I have left out the intermediate arrow\, making it less of an
> eyesore and easier to understand as a two dimensional hash\.
>
> coralament / best Grötens / liebe Grüße / best regards / elkorajn salutojn
> Daniel Pfeiffer

**Comment by p5pRT** [comment]
### From @jkeenan

This ticket appears to have been created in anticipation of receipt of a
patch\, but none has appeared\.

Can the ticket be closed?

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @jkeenan

On Sun Nov 20 09&#8203;:03&#8203;:34 2011\, jkeenan wrote&#8203;:
> This ticket appears to have been created in anticipation of receipt of a
> patch\, but none has appeared\.
> 
> Can the ticket be closed?

This ticket appears to duplicate
https://rt-archive.perl.org/perl5/Ticket/Display.html?id=74650.

It should be merged into that one\.

Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
@jkeenan - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @jkeenan

On Sun Apr 25 06&#8203;:21&#8203;:03 2010\, occitan@&#8203;esperanto\.org wrote&#8203;:
> In the dereference benchmark you make exactly the mistake you warn of 
> before&#8203;: you change several factors at once and show an overdrawn 
> conclusion\.  While refactoring out the common $ref\->\{ref\} to access it 
> only once\, you unnecessarily also copy both strings to a local 
> variable\.  It is this second action that drives the time up\, if you 
> leave that out and run timethese several times your findings are still 
> right\, but far less\.  If however you need a third or more values \(which 
> I added of the same length\)\, dereferencing the outer hash only once 
> becomes a clear winner&#8203;:
> 
>     timethese\(1000000\, \{
>         '2direct'      => sub \{
>             my $x = $ref\->\{ref\}\{\_myscore\} \. $ref\->\{ref\}\{\_yourscore\} ;
>         \}\,
>         '2wrong'       => sub \{
>             my $ref = $ref\->\{ref\};
>             my $myscore = $ref\->\{\_myscore\};        \# why?
>             my $yourscore = $ref\->\{\_yourscore\};    \# why?
>             my $x = $myscore \. $yourscore;
>         \}\,
>         '2dereference' => sub \{
>             my $ref = $ref\->\{ref\};
>             my $x = $ref\->\{\_myscore\} \. $ref\->\{\_yourscore\};
>         \}\,
>         '3direct'      => sub \{
>             my $x = $ref\->\{ref\}\{\_myscore\} \. $ref\->\{ref\}\{\_yourscore\} \.
>                 $ref\->\{ref\}\{\_ourscore\};
>         \}\,
>         '3dereference' => sub \{
>             my $ref = $ref\->\{ref\};
>             my $x = $ref\->\{\_myscore\} \. $ref\->\{\_yourscore\} \.
>                 $ref\->\{\_ourscore\};

I get "Use of uninitialized value in concatenation \(\.\) or string" on the
line above\.

>         \}\,
>     \}\);
> 
> Btw\. I have left out the intermediate arrow\, making it less of an 
> eyesore and easier to understand as a two dimensional hash\.
> 
> coralament / best Gr�tens / liebe Gr��e / best regards / elkorajn salutojn
> Daniel Pfeiffer
> 


Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
### From occitan@esperanto.org

la 2013\-08\-10 03&#8203;:36 James E Keenan via RT skribis&#8203;:
> On Sun Apr 25 06&#8203;:21&#8203;:03 2010\, occitan@&#8203;esperanto\.org wrote&#8203;:
>> In the dereference benchmark you make exactly the mistake you warn of 
>> before&#8203;: you change several factors at once and show an overdrawn 
>> conclusion\.  While refactoring out the common $ref\->\{ref\} to access it 
>> only once\, you unnecessarily also copy both strings to a local 
>> variable\.  It is this second action that drives the time up\, if you 
>> leave that out and run timethese several times your findings are still 
>> right\, but far less\.  If however you need a third or more values \(which 
>> I added of the same length\)\, dereferencing the outer hash only once 
>> becomes a clear winner&#8203;:
>>
>>     timethese\(1000000\, \{
>>         '2direct'      => sub \{
>>             my $x = $ref\->\{ref\}\{\_myscore\} \. $ref\->\{ref\}\{\_yourscore\} ;
>>         \}\,
>>         '2wrong'       => sub \{
>>             my $ref = $ref\->\{ref\};
>>             my $myscore = $ref\->\{\_myscore\};        \# why?
>>             my $yourscore = $ref\->\{\_yourscore\};    \# why?
>>             my $x = $myscore \. $yourscore;
>>         \}\,
>>         '2dereference' => sub \{
>>             my $ref = $ref\->\{ref\};
>>             my $x = $ref\->\{\_myscore\} \. $ref\->\{\_yourscore\};
>>         \}\,
>>         '3direct'      => sub \{
>>             my $x = $ref\->\{ref\}\{\_myscore\} \. $ref\->\{ref\}\{\_yourscore\} \.
>>                 $ref\->\{ref\}\{\_ourscore\};
>>         \}\,
>>         '3dereference' => sub \{
>>             my $ref = $ref\->\{ref\};
>>             my $x = $ref\->\{\_myscore\} \. $ref\->\{\_yourscore\} \.
>>                 $ref\->\{\_ourscore\};
> I get "Use of uninitialized value in concatenation \(\.\) or string" on the
> line above\.
>
>>         \}\,
>>     \}\);
>>
>> Btw\. I have left out the intermediate arrow\, making it less of an 
>> eyesore and easier to understand as a two dimensional hash\.
>>
>> Thank you very much\.
>> Jim Keenan
>>

You'd obvoiusly have to add \_ouscore to the inital hash\.  I'd also look at
taking a hash slice and joining that\, even though I fear it might be slower
\(and somewhat messy syntactically\.

안녕히 계세요 / coralament / best Grötens / liebe Grüße / best regards / elkorajn salutojn
Daniel Pfeiffer

\-\- 
배운다 / lerne / learn / apprends  Esperanto&#8203;:
&nbsp;                    http&#8203;://lernu\.net  /  http&#8203;://ikurso\.net
Reliability\, Perl programming and much more in Makefiles&#8203;:
&nbsp;                    http&#8203;://makepp\.sourceforge\.net

**Comment by p5pRT** [comment]
### From @jkeenan

On Sat Aug 10 09&#8203;:48&#8203;:14 2013\, occitan@&#8203;esperanto\.org wrote&#8203;:
> la 2013\-08\-10 03&#8203;:36 James E Keenan via RT skribis&#8203;:
> > On Sun Apr 25 06&#8203;:21&#8203;:03 2010\, occitan@&#8203;esperanto\.org wrote&#8203;:
> >> In the dereference benchmark you make exactly the mistake you warn
> of
> >> before&#8203;: you change several factors at once and show an overdrawn
> >> conclusion\.  While refactoring out the common $ref\->\{ref\} to access
> it
> >> only once\, you unnecessarily also copy both strings to a local
> >> variable\.  It is this second action that drives the time up\, if you
> >> leave that out and run timethese several times your findings are
> still
> >> right\, but far less\.  If however you need a third or more values
> \(which
> >> I added of the same length\)\, dereferencing the outer hash only once
> >> becomes a clear winner&#8203;:
> >>
> >>     timethese\(1000000\, \{
> >>         '2direct'      => sub \{
> >>             my $x = $ref\->\{ref\}\{\_myscore\} \. $ref\->\{ref\}\{\_yourscore\}
> ;
> >>         \}\,
> >>         '2wrong'       => sub \{
> >>             my $ref = $ref\->\{ref\};
> >>             my $myscore = $ref\->\{\_myscore\};        \# why?
> >>             my $yourscore = $ref\->\{\_yourscore\};    \# why?
> >>             my $x = $myscore \. $yourscore;
> >>         \}\,
> >>         '2dereference' => sub \{
> >>             my $ref = $ref\->\{ref\};
> >>             my $x = $ref\->\{\_myscore\} \. $ref\->\{\_yourscore\};
> >>         \}\,
> >>         '3direct'      => sub \{
> >>             my $x = $ref\->\{ref\}\{\_myscore\} \. $ref\->\{ref\}\{\_yourscore\}
> \.
> >>                 $ref\->\{ref\}\{\_ourscore\};
> >>         \}\,
> >>         '3dereference' => sub \{
> >>             my $ref = $ref\->\{ref\};
> >>             my $x = $ref\->\{\_myscore\} \. $ref\->\{\_yourscore\} \.
> >>                 $ref\->\{\_ourscore\};
> > I get "Use of uninitialized value in concatenation \(\.\) or string" on
> the
> > line above\.
> >
> >>         \}\,
> >>     \}\);
> >>
> >> Btw\. I have left out the intermediate arrow\, making it less of an
> >> eyesore and easier to understand as a two dimensional hash\.
> >>
> >> Thank you very much\.
> >> Jim Keenan
> >>
> 
> You'd obvoiusly have to add \_ouscore to the inital hash\.  I'd also
> look at
> taking a hash slice and joining that\, even though I fear it might be
> slower
> \(and somewhat messy syntactically\.
> 
> 안녕히 계세요 / coralament / best Grötens / liebe Grüße / best regards /
> elkorajn salutojn
> Daniel Pfeiffer
> 


I have to admit that I'm not comfortable being asked to edit a patch
originally submitted more than three years ago\.  I think we need
additional eyeballs on Daniel's original critique of pod/perlperf\.pod\. 
Is the code example incorrect?  If so\, how do we improve it?

Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
### From occitan@esperanto.org

la 2013\-08\-11 06&#8203;:26 James E Keenan via RT skribis&#8203;:
> I have to admit that I'm not comfortable being asked to edit a patch
> originally submitted more than three years ago\. I think we need additional
> eyeballs on Daniel's original critique of pod/perlperf\.pod\. Is the code
> example incorrect? If so\, how do we improve it? Thank you very much\. Jim Keenan 

Ok\, I wondered about the join slice variant\, so I've added it\, but it performs
about as badly \(once faster\, twice slower\) as the wrong code in perlperf\. 
I've also had to bump it to 10 million\, to take modern computing power into
account\.

My criticism was and still is that the shown test doesn't test what it
pretends to\.  It talks about dereferencing strategies\, but then shows a far
bigger performance loss\, which instead comes from copying the 2 strings to
my\-variables\.  That has nothing to do with the announced measurement\, and
completely distorts it\!  Therefore I labelled it wrong and want it out of the doc\.

And I have the three\-element variants for comparison\.  Because the whole test
is about a tradeoff between an extra copying operation versus dereferencing
for each access\.  As tradeoffs go\, they depend heavily on the proportion\. 
When there are two accesses\, the extra copy outweighs by about 3%\, but at
three accesses it is already break even\.  That's because the extra copy is
constantly done once\.

안녕히 계세요 / coralament / best Grötens / liebe Grüße / best regards / elkorajn salutojn
Daniel Pfeiffer

\-\- 
배운다 / lerne / learn / apprends  Esperanto&#8203;:
&nbsp;                    http&#8203;://lernu\.net  /  http&#8203;://ikurso\.net
Reliability\, Perl programming and much more in Makefiles&#8203;:
&nbsp;                    http&#8203;://makepp\.sourceforge\.net

**Comment by p5pRT** [comment]
### From occitan@esperanto.org

[timetest.pl](https://rt-archive.perl.org/perl5/Ticket/Attachment/1243389/648629/timetest.pl)

**Comment by p5pRT** [comment]
### From @richardfoley

It's good to see this being looked at\, and updated appropriately\. Good job\.

\-\- 
Ciao

Richard Foley

http&#8203;://www\.rfi\.net/books\.html

On Sun\, Aug 11\, 2013 at 05&#8203;:26&#8203;:28PM \+0200\, Daniel Pfeiffer wrote&#8203;:
> la 2013\-08\-11 06&#8203;:26 James E Keenan via RT skribis&#8203;:
> > I have to admit that I'm not comfortable being asked to edit a patch
> > originally submitted more than three years ago\. I think we need additional
> > eyeballs on Daniel's original critique of pod/perlperf\.pod\. Is the code
> > example incorrect? If so\, how do we improve it? Thank you very much\. Jim Keenan 
> 
> Ok\, I wondered about the join slice variant\, so I've added it\, but it performs
> about as badly \(once faster\, twice slower\) as the wrong code in perlperf\. 
> I've also had to bump it to 10 million\, to take modern computing power into
> account\.
> 
> My criticism was and still is that the shown test doesn't test what it
> pretends to\.  It talks about dereferencing strategies\, but then shows a far
> bigger performance loss\, which instead comes from copying the 2 strings to
> my\-variables\.  That has nothing to do with the announced measurement\, and
> completely distorts it\!  Therefore I labelled it wrong and want it out of the doc\.
> 
> And I have the three\-element variants for comparison\.  Because the whole test
> is about a tradeoff between an extra copying operation versus dereferencing
> for each access\.  As tradeoffs go\, they depend heavily on the proportion\. 
> When there are two accesses\, the extra copy outweighs by about 3%\, but at
> three accesses it is already break even\.  That's because the extra copy is
> constantly done once\.
> 
> 안녕히 계세요 / coralament / best Grötens / liebe Grüße / best regards / elkorajn salutojn
> Daniel Pfeiffer
> 
> \-\- 
> 배운다 / lerne / learn / apprends  Esperanto&#8203;:
>                     http&#8203;://lernu\.net  /  http&#8203;://ikurso\.net
> Reliability\, Perl programming and much more in Makefiles&#8203;:
>                     http&#8203;://makepp\.sourceforge\.net
>

**Comment by p5pRT** [comment]
### From @tonycoz

On Mon Aug 12 02&#8203;:09&#8203;:42 2013\, richardfoley wrote&#8203;:
> It's good to see this being looked at\, and updated appropriately\. Good job\.
> 


Attached is an actual patch the modifies the dereference benchmark to avoid the
extra copying objected to by the OP and updates the benchmark results\.

Tony

**Comment by p5pRT** [comment]
### From @tonycoz

<details><summary>0001-perl-74650-avoid-some-copying-for-a-more-realistic-c.patch</summary>

```diff
From 49d4fbbc7a57d87fb1d60418a3c0e7e1cad971a4 Mon Sep 17 00:00:00 2001
From: Tony Cook <tony@develop-help.com>
Date: Thu, 28 Nov 2013 16:16:48 +1100
Subject: [PATCH] [perl #74650] avoid some copying for a more realistic change
 to benchmark

Also, update the benchmark results to match.

Benchmarked on a mostly idle Mac Mini.
---
 pod/perlperf.pod |   16 +++++++---------
 1 file changed, 7 insertions(+), 9 deletions(-)

diff --git a/pod/perlperf.pod b/pod/perlperf.pod
index 007a02b..9e441e7 100644
--- a/pod/perlperf.pod
+++ b/pod/perlperf.pod
@@ -138,15 +138,13 @@ comparative code in a file and running a C<Benchmark> test.
             },
     };
 
-    timethese(1000000, {
+    timethese(10000000, {
             'direct'       => sub {
                 my $x = $ref->{ref}->{_myscore} . $ref->{ref}->{_yourscore} ;
             },
             'dereference'  => sub {
                 my $ref  = $ref->{ref};
-                my $myscore = $ref->{_myscore};
-                my $yourscore = $ref->{_yourscore};
-                my $x = $myscore . $yourscore;
+                my $x = $ref->{_myscore} . $ref->{_yourscore};
             },
     });
 
@@ -159,13 +157,13 @@ output by the C<Benchmark> module, to see which approach is the most effective.
 
     $> perl dereference
 
-    Benchmark: timing 1000000 iterations of dereference, direct...
-    dereference:  2 wallclock secs ( 1.59 usr +  0.00 sys =  1.59 CPU) @ 628930.82/s (n=1000000)
-        direct:  1 wallclock secs ( 1.20 usr +  0.00 sys =  1.20 CPU) @ 833333.33/s (n=1000000)
+    Benchmark: timing 10000000 iterations of dereference, direct...
+    dereference:  5 wallclock secs ( 4.76 usr +  0.00 sys =  4.76 CPU) @ 2100840.34/s (n=10000000)
+        direct:  4 wallclock secs ( 4.14 usr +  0.00 sys =  4.14 CPU) @ 2415458.94/s (n=10000000)
 
 The difference is clear to see and the dereferencing approach is slower.  While
-it managed to execute an average of 628,930 times a second during our test, the
-direct approach managed to run an additional 204,403 times, unfortunately.
+it managed to execute an average of 2,100,840 times a second during our test, the
+direct approach managed to run an additional 314,618 times, unfortunately.
 Unfortunately, because there are many examples of code written using the
 multiple layer direct variable access, and it's usually horrible.  It is,
 however, minusculy faster.  The question remains whether the minute gain is
-- 
1.7.10.4


```
</details>

**Comment by p5pRT** [comment]
### From @jkeenan

On Wed Nov 27 21&#8203;:18&#8203;:17 2013\, tonyc wrote&#8203;:
> On Mon Aug 12 02&#8203;:09&#8203;:42 2013\, richardfoley wrote&#8203;:
> > It's good to see this being looked at\, and updated appropriately\.
> > Good job\.
> >
> 
> 
> Attached is an actual patch the modifies the dereference benchmark to
> avoid the
> extra copying objected to by the OP and updates the benchmark results\.
> 
> Tony


Tony\, I think your patch is good\.  However\, I've always felt that the existing language in the final paragraph \-\- the one evaluating the benchmarking results \-\- is verbose and confusing\.

I'm attaching a patch which trims that paragraph down\.  I used the existing documentation's numbers; we'll have to replace those once we finalize the explanatory language\.

Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
### From @jkeenan

<details><summary>0001-Simplify-discussion-of-benchmarking-results.patch</summary>

```diff
From 93f61b95f5e31c5c4256675f770d35ca164d1ef7 Mon Sep 17 00:00:00 2001
From: James E Keenan <jkeenan@cpan.org>
Date: Thu, 28 Nov 2013 15:07:24 +0100
Subject: [PATCH] Simplify discussion of benchmarking results.

---
 pod/perlperf.pod |   11 ++++-------
 1 files changed, 4 insertions(+), 7 deletions(-)

diff --git a/pod/perlperf.pod b/pod/perlperf.pod
index 007a02b..b6cb2ca 100644
--- a/pod/perlperf.pod
+++ b/pod/perlperf.pod
@@ -163,13 +163,10 @@ output by the C<Benchmark> module, to see which approach is the most effective.
     dereference:  2 wallclock secs ( 1.59 usr +  0.00 sys =  1.59 CPU) @ 628930.82/s (n=1000000)
         direct:  1 wallclock secs ( 1.20 usr +  0.00 sys =  1.20 CPU) @ 833333.33/s (n=1000000)
 
-The difference is clear to see and the dereferencing approach is slower.  While
-it managed to execute an average of 628,930 times a second during our test, the
-direct approach managed to run an additional 204,403 times, unfortunately.
-Unfortunately, because there are many examples of code written using the
-multiple layer direct variable access, and it's usually horrible.  It is,
-however, minusculy faster.  The question remains whether the minute gain is
-actually worth the eyestrain, or the loss of maintainability.
+The difference is clear to see: the C<direct> approach is faster.  C<direct>
+ran 204,403 times/second more than the C<dereference> approach.  The question
+remains, however, whether C<direct>'s relatively modest performance gain
+outweighs its poorer readability and maintainability.
 
 =head2  Search and replace or tr
 
-- 
1.7.1


```
</details>

**Comment by p5pRT** [comment]
### From @tonycoz

On Thu Nov 28 06&#8203;:14&#8203;:19 2013\, jkeenan wrote&#8203;:
> I'm attaching a patch which trims that paragraph down\.  I used the
> existing documentation's numbers; we'll have to replace those once we
> finalize the explanatory language\.

The main improvement I can see is some sort of relative measure \- the original
text presented the "additional 204\,403 times" and the "628\,930 timesa second" in the same sentence\, so you can see you have about a 30% improvement\.

Perhaps the new text should give the improvement as a percentage \(15% with the new benchmark\)\.

Tony

**Comment by p5pRT** [comment]
### From @tonycoz

<details><summary>0002-Simplify-discussion-of-benchmarking-results.patch</summary>

```diff
From 6d4d14cb6636dfa6b92b290637ed0ae6e8d480b6 Mon Sep 17 00:00:00 2001
From: James E Keenan <jkeenan@cpan.org>
Date: Thu, 28 Nov 2013 15:07:24 +0100
Subject: [PATCH 2/2] Simplify discussion of benchmarking results.

---
 pod/perlperf.pod |   11 ++++-------
 1 file changed, 4 insertions(+), 7 deletions(-)

diff --git a/pod/perlperf.pod b/pod/perlperf.pod
index 9e441e7..7774aac 100644
--- a/pod/perlperf.pod
+++ b/pod/perlperf.pod
@@ -161,13 +161,10 @@ output by the C<Benchmark> module, to see which approach is the most effective.
     dereference:  5 wallclock secs ( 4.76 usr +  0.00 sys =  4.76 CPU) @ 2100840.34/s (n=10000000)
         direct:  4 wallclock secs ( 4.14 usr +  0.00 sys =  4.14 CPU) @ 2415458.94/s (n=10000000)
 
-The difference is clear to see and the dereferencing approach is slower.  While
-it managed to execute an average of 2,100,840 times a second during our test, the
-direct approach managed to run an additional 314,618 times, unfortunately.
-Unfortunately, because there are many examples of code written using the
-multiple layer direct variable access, and it's usually horrible.  It is,
-however, minusculy faster.  The question remains whether the minute gain is
-actually worth the eyestrain, or the loss of maintainability.
+The difference is clear to see: the C<direct> approach is faster.  C<direct>
+ran 314,618 times/second more than the C<dereference> approach.  The question
+remains, however, whether C<direct>'s relatively modest performance gain
+outweighs its poorer readability and maintainability.
 
 =head2  Search and replace or tr
 
-- 
1.7.10.4


```
</details>

**Comment by p5pRT** [comment]
@dcollinsn - Status changed from 'open' to 'stalled'

**Comment by khwilliamson** [comment]
Why not just apply @tonycoz 's patch?

**Comment by tonycoz** [comment]
I'll look over it some time today and make a PR.
